bench: update logging benchmarks

jonatack opened the pull request:
Update our logging benchmarks for evaluating ongoing work like #25203 and refactoring proposals like #26619 and #26697.

- make the logging benchmarks order-independent (Larry Ruane)
- add missing benchmarks for the `LogPrintLevel` and `LogPrintfCategory` macros that our logging is migrating to; at some later point it should be feasible to drop some of the previous logging benchmarks
- update the logging benchmark naming to be clear which benchmark corresponds to which log macro, and update the ordering to be the same as the output
- add clarifying documentation to the logging benchmarks
- improve the `-debuglogfile` config option help to be clearer; can be tested by running `./src/bitcoind -help | grep -A4 '\-debuglogfile'`

Reviewers can run the logging benchmarks with:
```bash
./src/bench/bench_bitcoin -filter='LogP*.*'
```

jonatack committed:
bench: add LogPrintfCategory and LogPrintLevel benchmarks

for these new macros that our logging is planned to migrate to.  At some
point it may be feasible to drop some of the previous logging benchmarks.

jonatack committed:
bench: update logging benchmark naming for clarity

to better track which benchmark corresponds to which log macro.

jonatack committed:
bench: order the logging benchmark code by output

jonatack committed:
bench: document expected results in logging benchmarks

and clarify the intention behind the -nodebuglogfile bench.

Co-authored-by: "kouloumos <kouloumosa@gmail.com>"
Co-authored-by: "Larry Ruane <larryruane@gmail.com>"

kouloumos commented on src/bench/logging.cpp:
```diff
@@ -17,32 +17,62 @@ static void Logging(benchmark::Bench& bench, const std::vector<const char*>& ext
+        LogPrint(BCLog::VALIDATION, "%s\n", "test");
+        LogPrintfCategory(BCLog::VALIDATION, "%s\n", "test");
         LogPrintf("%s\n", "test");
-        LogPrint(BCLog::NET, "%s\n", "test");
```
The way this benchmark is implemented, we can identify a regression between code changes. But if we wanted to get a sense of how logging without a a debug.log file compares to logging with a debug.log file, this doesn't help, right? It should only log with one of {LogPrintLevel, LogPrint, LogPrintfCategory} in order to compare the result with one of the other benchmarks.

jonatack replied:
That benchmark was added in fa31dc9b714 with the intention expressed in https://github.com/bitcoin/bitcoin/pull/18815#issuecomment-621404493, i.e. to test this code currently in `setup_common`:

```cpp
/**
 * Make a test setup that has disk access to the debug.log file disabled. Can
 * be used in "hot loops", for example fuzzing or benchmarking.
 */
template <class T = const BasicTestingSetup>
std::unique_ptr<T> MakeNoLogFileContext(const std::string& chain_name = CBaseChainParams::REGTEST, const std::vector<const char*>& extra_args = {})
{
    const std::vector<const char*> arguments = Cat(
        {
            "-nodebuglogfile",
            "-nodebug",
        },
        extra_args);

    return std::make_unique<T>(chain_name, arguments);
}
```
So it looks like the benchmark should be updated to pass `-nodebug` (and still bench with all macros passed here just in case). It also looks like the -debuglogfile help could be clearer. Updating.

jonatack replied:
Thanks @kouloumos. Decided to add a comment to clarify the intention behind that -nodebuglogfile bench, and not add macros to it per your suggestion.

jonatack committed:
doc: improve -debuglogfile help to be a bit clearer

kouloumos commented on src/bench/logging.cpp:
```diff
@@ -17,32 +17,61 @@ static void Logging(benchmark::Bench& bench, const std::vector<const char*>& ext
+
+static void LogPrintWithCategory(benchmark::Bench& bench)
+{
+    Logging(bench, {"-logthreadnames=0", "-debug=net"}, [] { LogPrint(BCLog::VALIDATION, "%s\n", "test"); });
```
What's the expected behavior when logging with a category that is not specified in `-debug=`?

Does logging with `BCLog::VALIDATION` but with `-debug=net` means that `BCLog::Logger::WillLogCategoryLevel` will return `false`? Therefore `LogPrint()` will not actually log?

LarryRuane replied:
> What's the expected behavior when logging with a category that is not specified in -debug=?

It should not log. Note that the `LogPrintWithoutCategory` benchmark specifies `-debug=0` which disables all categories, although this has a problem that I mentioned in my other comment.

> Does logging with BCLog::VALIDATION but with -debug=net means that BCLog::Logger::WillLogCategoryLevel will return false? Therefore LogPrint() will not actually log?

That's correct.

kouloumos replied:
This is different from what I observe running this benchmark.

When `BCLog::NET` was replaced with `BCLog::VALIDATION` it seems that changing `-debug=net` to `-debug=validation` was overlooked. Which per my understanding, means that `LogPrintWithCategory` should not actually benchmark what is supposed to, but rather be comparable to `LogPrintWithoutCategory`. This is not true and I only see the expected behavior when I change `-debug=net` to `-debugexclude=validation`.

I will investigate this further.

LarryRuane replied:
I think I figured out this mystery. It's a wild story. I'm almost certain you're right that the test has a mistake in it:
```
Logging(bench, {"-logthreadnames=0", "-debug=net"}, [] { LogPrint(BCLog::VALIDATION, "%s\n", "test"); });
```
This is meant to specify the _same_ category (not the different categories, "net" and `VALIDATION`). But the confusion is that with this mistake, the benchmark still shows a large `ns/op`, as if it's actually logging. I ran this code with the debugger, which shows that it is indeed logging the message. The reason, I just figured out, is that the test framework [enables _all_ logging](https://github.com/bitcoin/bitcoin/blob/7241b936c5b4d004cdbdc7dc4d0b586856f6716b/src/test/util/setup_common.cpp#L113), so when the test specifies `-debug=net`, that has no effect because all categories are enabled anyway. 

I tried changing that line of benchmark code to:
```
Logging(bench, {"-logthreadnames=0", "-debug=0", "-debug=net"}, [] { LogPrint(BCLog::VALIDATION, "%s\n", "test"); });
```
Note the addition of `-debug=0` which should disable all logging. So now what should happen is, the test framework specifies `-debug` which enables all categories (the test framework arguments occur before the ones we specify here), but then we specify `-debug=0` which _disables_ all categories, then add back just the one we want. Now, no logging should actually happen, since we haven't enabled `"validation"`. And this was the case, no logging occurred (the `ns/op` was a low number).

Now, let's just fix the bug in the test:
```
Logging(bench, {"-logthreadnames=0", "-debug=0", "-debug=validation"}, [] { LogPrint(BCLog::VALIDATION, "%s\n", "test"); });
```
To my surprise, this also had a low `ns/op`, meaning that it was not logging! How could that be, when we're explicitly specifying `-debug=validation` and we're doing a `VALIDATION` print?

Further investigation revealed that when the command line includes `-debug=0`, then we [_do not_ enable](https://github.com/bitcoin/bitcoin/blob/7241b936c5b4d004cdbdc7dc4d0b586856f6716b/src/init/common.cpp#L86) _any categories_ specified on the command line! That is, the `-debug=0` causes even later `-debug=category` arguments to be ignored!

Anyway, I believe this is the fix we want, this gives the expected results (actual logging, a high `ns/op`):
```diff
 static void LogPrintWithCategory(benchmark::Bench& bench)
 {
-    Logging(bench, {"-logthreadnames=0", "-debug=net"}, [] { LogPrint(BCLog::VALIDATION, "%s\n", "test"); });
+     // the test framework enables all categories by default
+    Logging(bench, {"-logthreadnames=0"}, [] { LogPrint(BCLog::VALIDATION, "%s\n", "test"); });
 }
```

kouloumos reviewed (commented):
Concept ACK
The addition of the new macros makes sense as well as the new names for the benchmarks. 

Just one observation: although I initially was sceptical about the {No,Yo} naming scheme, looking at the updated {With,Without} it seems that is now harder to skim through the list.

LarryRuane commented on src/bench/logging.cpp:
```diff
@@ -17,32 +17,61 @@ static void Logging(benchmark::Bench& bench, const std::vector<const char*>& ext
+    Logging(bench, {"-logthreadnames=0", "-debug=net"}, [] { LogPrint(BCLog::VALIDATION, "%s\n", "test"); });
+}
+
+static void LogPrintWithoutCategory(benchmark::Bench& bench)
```
@kouloumos I'm seeing the same thing. After investigation, I found that the problem is that the global logging object instance is not re-created for each run. When `LogPrintWithoutCategory` runs, it still has logging categories enabled from the previous benchmark runs.

This patch fixes the problem for me, give it a try. The change to `LogInstance()` looks racy (i.e. what if two threads both see `g_logger` is null, and they both create a new object), but this change is safe because `LogInstance()` first runs as part of the [global static initialization](https://en.cppreference.com/w/cpp/language/initialization), before `main()` is called, so there's only one thread. But if there's concern about this, `std::call_once()` can be used (it's used in a few places already).

<details>
<summary>patch</summary>

```diff
diff --git a/src/bench/logging.cpp b/src/bench/logging.cpp
index d5c78932bc..434d1c58d4 100644
--- a/src/bench/logging.cpp
+++ b/src/bench/logging.cpp
@@ -9,6 +9,7 @@
 
 static void Logging(benchmark::Bench& bench, const std::vector<const char*>& extra_args, const std::function<void()>& log)
 {
+    LogReset();
     TestingSetup test_setup{
         CBaseChainParams::REGTEST,
         extra_args,
diff --git a/src/logging.cpp b/src/logging.cpp
index 8bf60169b0..24f3338368 100644
--- a/src/logging.cpp
+++ b/src/logging.cpp
@@ -18,6 +18,8 @@
 const char * const DEFAULT_DEBUGLOGFILE = "debug.log";
 constexpr auto MAX_USER_SETABLE_SEVERITY_LEVEL{BCLog::Level::Info};
 
+static BCLog::Logger* g_logger;
+
 BCLog::Logger& LogInstance()
 {
 /**
@@ -35,10 +37,16 @@ BCLog::Logger& LogInstance()
  * This method of initialization was originally introduced in
  * ee3374234c60aba2cc4c5cd5cac1c0aefc2d817c.
  */
-    static BCLog::Logger* g_logger{new BCLog::Logger()};
+    if (g_logger == nullptr) g_logger = new BCLog::Logger();
     return *g_logger;
 }
 
+void LogReset()
+{
+    delete g_logger;
+    g_logger = new BCLog::Logger();
+}
+
 bool fLogIPs = DEFAULT_LOGIPS;
 
 static int FileWriteStr(const std::string &str, FILE *fp)
diff --git a/src/logging.h b/src/logging.h
index 16f23a19e8..522923c000 100644
--- a/src/logging.h
+++ b/src/logging.h
@@ -266,6 +266,7 @@ namespace BCLog {
 } // namespace BCLog
 
 BCLog::Logger& LogInstance();
+void LogReset(); // testing only
 
 /** Return true if log accepts specified category, at the specified level. */
 static inline bool LogAcceptCategory(BCLog::LogFlags category, BCLog::Level level)
```
</details>

LarryRuane committed:
bench: allow logging benchmarks to be order-independent

The global logging object instance is not re-created for each run, so when
multiple logging benchmarks are run, each one after the first one still has
the logging categories enabled from the previous ones.  This commit disables
all categories at the start of each benchmark.

LarryRuane reviewed (commented):
ACK a0e092e1af6f8903568d48ebb04ef654a6b71d53
```
$ ./src/bench/bench_bitcoin -filter='LogP*.*'
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|            2,716.93 |          368,062.78 |    0.5% |      0.01 | `LogPrintLevelWithThreadNames`
|            2,691.57 |          371,529.69 |    0.4% |      0.01 | `LogPrintLevelWithoutThreadNames`
|            2,657.64 |          376,273.25 |    0.3% |      0.01 | `LogPrintWithCategory`
|                4.15 |      241,089,995.32 |    0.3% |      0.01 | `LogPrintWithoutCategory`
|            2,705.19 |          369,659.92 |    0.2% |      0.01 | `LogPrintfCategoryWithThreadNames`
|            2,651.54 |          377,138.94 |    0.1% |      0.01 | `LogPrintfCategoryWithoutThreadNames`
|            2,649.92 |          377,370.03 |    0.3% |      0.01 | `LogPrintfWithThreadNames`
|            2,613.43 |          382,638.59 |    0.3% |      0.01 | `LogPrintfWithoutThreadNames`
|               48.00 |       20,832,403.39 |    0.3% |      0.01 | `LogWithoutWriteToFile`
```
I believe these are the (rough) results we expect. If you retouch, consider adding a comment at the top of the file probably, stating the expected results -- that is, that all but two of the tests should have similar performance, but that `LogPrintWithoutCategory` should be about 3 orders of magnitude faster than most of the tests, and `LogWithoutWriteToFile` should be about 2 orders of magnitude faster than most of the tests. Maybe even explain briefly why -- for example, `LogWithoutWriteToFile` is still logging somewhere (?) but not to a file, which is why it's faster, while `LogPrintWithoutCategory` should be almost a no-op. Something like that, just so when people run this benchmark, they'll know roughly what's expected without having to dig in and really understand the code.

fanquake commented on src/bench/logging.cpp:
```diff
@@ -17,32 +19,77 @@ static void Logging(benchmark::Bench& bench, const std::vector<const char*>& ext
+    });
+}
+
+static void LogPrintfCategoryWithThreadNames(benchmark::Bench& bench)
```
In 4db9dc8ac3baa9751702ca575965661e63384692:

Looks like you're introducing these benchmarks, with `BCLog::NET`, only to change them in a later commit (dfb40748911b9553a6e10c8376b032dac36bf938) to `BCLog::VALIDATION`? It's not obvious to me why the category needs changing, but this can probably be re-ordered so we aren't changing the same (newly introduced) lines of code, multiple times in the same PR.

LarryRuane commented on src/logging.h:
```diff
@@ -201,6 +201,7 @@ namespace BCLog {
 } // namespace BCLog
 
 BCLog::Logger& LogInstance();
+void LogReset(); // for benchmarks and testing
```
Resetting categories would also touch real code, wouldn't it?

LarryRuane replied:
Thanks, @MarcoFalke, good suggestion, [this branch](https://github.com/LarryRuane/bitcoin/commits/pr26957-1) changes that commit to not touch production code ([individual commit](https://github.com/LarryRuane/bitcoin/commit/a1cd3cc17b9fdce1d60d662c447f845a12063d81)) @jonatack, feel free to cherry-pick.

LarryRuane commented:
ACK 8c47d599b87d6b2d43e7d37ce0aaf4f541535bb9
The commits are very clean.

fanquake commented on src/bench/logging.cpp:
```diff
@@ -17,32 +25,67 @@ static void Logging(benchmark::Bench& bench, const std::vector<const char*>& ext
+    });
+}
+
+static void LogPrintfWithThreadNames(benchmark::Bench& bench)
```
In 102b2033493f0d61e9763d094cb8a0017f7e3a10 bench: order the logging benchmark code by output: 

Given that the code-movement, and re-ordering the `BENCHMARK(` calls doesn't change the output, this seems like a no-op?

In general, is (re-)ordering code in a `.cpp` file based on the order functions are called something we do?

martinus replied:
Benchmarks are registered into a `std::map` so they are alphabetically sorted anyways https://github.com/bitcoin/bitcoin/blob/master/src/bench/bench.h#L68

martinus commented:
code review & tested ACK, here are my benchmark results: 

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|            3,822.12 |          261,634.61 |    0.3% |       22,931.00 |        8,536.02 |  2.686 |       4,512.00 |    0.1% |      0.53 | `LogPrintLevelWithThreadNames`
|            3,776.34 |          264,807.01 |    0.4% |       22,494.01 |        8,370.09 |  2.687 |       4,403.00 |    0.1% |      0.48 | `LogPrintLevelWithoutThreadNames`
|            3,733.48 |          267,846.56 |    0.5% |       22,252.00 |        8,272.50 |  2.690 |       4,351.00 |    0.1% |      0.54 | `LogPrintWithCategory`
|                7.25 |      137,992,883.01 |    0.5% |           71.00 |           23.04 |  3.081 |          16.00 |    0.0% |      0.67 | `LogPrintWithoutCategory`
|            3,779.40 |          264,592.05 |    0.3% |       22,773.01 |        8,416.90 |  2.706 |       4,473.00 |    0.1% |      0.54 | `LogPrintfCategoryWithThreadNames`
|            3,709.16 |          269,602.50 |    0.4% |       22,122.00 |        8,208.57 |  2.695 |       4,318.00 |    0.1% |      0.53 | `LogPrintfCategoryWithoutThreadNames`
|            3,737.44 |          267,562.59 |    0.4% |       22,495.01 |        8,290.13 |  2.713 |       4,399.00 |    0.1% |      0.53 | `LogPrintfWithThreadNames`
|            3,630.74 |          275,425.89 |    0.1% |       21,844.00 |        7,989.23 |  2.734 |       4,244.00 |    0.1% |      0.53 | `LogPrintfWithoutThreadNames`
|              100.84 |        9,916,479.93 |    0.0% |          802.00 |          320.90 |  2.499 |         186.00 |    0.0% |      0.55 | `LogWithoutWriteToFile`

By the way, here is a nice way to create & share detailed benchmark results, see the results here: https://share.firefox.dev/3Z02boN

How to do this:

1. Run a benchmark for a few seconds and record data with perf
   ```sh
   perf record --call-graph lbr ./src/bench/bench_bitcoin -filter="LogPrintLevelWithoutThreadNames" -min-time=5000
   ```
2. Convert into data readable with https://profiler.firefox.com/
   ```sh
   perf script -F +pid > /tmp/test.perf
   ```
3. Drag & drop `test.perf` into https://profiler.firefox.com/, upload the data, and create a permalink to it

LarryRuane commented:
Here's a suggestion about benchmarking in general, wonder what you think. The idea is a kind of performance regression testing.

If there are a set of related benchmarks, as there are here, there may be some expected performance relationship between them. (Above I [suggested](https://github.com/bitcoin/bitcoin/pull/26957#pullrequestreview-1314607947) that Jon add a comment describing this relationship, and he [did](https://github.com/bitcoin/bitcoin/pull/26957/files#diff-b65a90c310b14b523533c724ac297ed970168e01e9589352700c7c52353609f3R9-R13).) It would be cool if this could somehow be specified in the code and checked automatically, and a failure raised if the runtime ratio between the sub-benchmarks exceeded some threshold. So for example, if a bug was introduced that made `LogPrintWithoutCategory` run at about the same speed as the other types of logging, flag an error.

What do you think? (Obviously not for this PR, but as a separate PR.)

achow101 commented:
ACK 8c47d599b87d6b2d43e7d37ce0aaf4f541535bb9
